File: backend/app/lib/signal.py
"""Server-side buy signal computation — mirrors frontend/src/lib/signal.ts logic."""


def compute_buy_signal(stock: dict, all_stocks: list, ai_score: int = None) -> dict:
    reasons = []
    score = 0

    pct_change = stock.get("percent_change", 0)
    high = stock.get("high", 0)
    ltp = stock.get("ltp", 0)
    volume = stock.get("volume", 0)
    prev_close = stock.get("prev_close", 0)
    open_price = stock.get("open", 0)

    # RED FLAGS
    if pct_change > 15:
        return {"signal": "RED", "label": "Avoid", "reasons": ["Up >15% — too late, high reversal risk"]}

    if ai_score is not None and ai_score < 40:
        return {"signal": "RED", "label": "Avoid", "reasons": ["AI score below 40"]}

    pct_from_high = ((high - ltp) / high * 100) if high > 0 else 0
    if pct_from_high > 5:
        reasons.append("Price dropped >5% from day high")
        score -= 3

    # GREEN CRITERIA
    if 3 <= pct_change <= 12:
        reasons.append(f"Good day gain: +{pct_change:.1f}%")
        score += 2
    elif pct_change >= 2:
        reasons.append(f"Moderate gain: +{pct_change:.1f}%")
        score += 1
    elif pct_change > 0:
        reasons.append(f"Small gain: +{pct_change:.1f}%")

    if pct_from_high <= 2:
        reasons.append("Near day high — strong buying pressure")
        score += 2
    elif pct_from_high <= 4:
        reasons.append("Slightly off day high")
        score += 1

    if all_stocks:
        volumes = sorted([s.get("volume", 0) for s in all_stocks], reverse=True)
        top30_idx = max(0, int(len(volumes) * 0.3))
        if volume >= volumes[top30_idx] and volume > 0:
            reasons.append("High volume — top 30%")
            score += 2

    gap_pct = ((open_price - prev_close) / prev_close * 100) if prev_close > 0 else 0
    if gap_pct >= 2:
        reasons.append(f"Gap up opening: +{gap_pct:.1f}%")
        score += 1

    if ai_score is not None:
        if ai_score >= 75:
            reasons.append(f"Strong AI score: {ai_score}")
            score += 3
        elif ai_score >= 65:
            reasons.append(f"Good AI score: {ai_score}")
            score += 2
        elif ai_score >= 50:
            reasons.append(f"Moderate AI score: {ai_score}")
            score += 1
        else:
            reasons.append(f"Low AI score: {ai_score}")
            score -= 1

    if score >= 6:
        return {"signal": "GREEN", "label": "Strong Buy", "reasons": reasons}
    if score >= 4:
        return {"signal": "GREEN", "label": "Buy", "reasons": reasons}
    if score >= 2:
        return {"signal": "YELLOW", "label": "Watch", "reasons": reasons}
    if score >= 0:
        return {"signal": "GRAY", "label": "Neutral", "reasons": reasons}
    return {"signal": "RED", "label": "Avoid", "reasons": reasons}

File: frontend/src/lib/signal.ts
export type SignalType = "GREEN" | "YELLOW" | "RED" | "GRAY";

export interface BuySignal {
  signal: SignalType;
  label: string;
  reasons: string[];
}

interface StockData {
  symbol: string;
  ltp: number;
  change: number;
  percent_change: number;
  open: number;
  high: number;
  low: number;
  prev_close: number;
  volume: number;
}

export function computeBuySignal(
  stock: StockData,
  allStocks: StockData[],
  aiScore?: number
): BuySignal {
  const reasons: string[] = [];
  let score = 0; // internal points system

  const pctChange = stock.percent_change;
  const high = stock.high;
  const ltp = stock.ltp;
  const volume = stock.volume;

  // --- RED FLAGS (instant avoid) ---

  if (pctChange > 15) {
    return {
      signal: "RED",
      label: "Avoid",
      reasons: ["Up >15% — too late, high reversal risk"],
    };
  }

  if (aiScore !== undefined && aiScore < 40) {
    return {
      signal: "RED",
      label: "Avoid",
      reasons: ["AI score below 40 — weak fundamentals or suspicious catalyst"],
    };
  }

  // Price fading from day high (momentum dying)
  const pctFromHigh = high > 0 ? ((high - ltp) / high) * 100 : 0;
  if (pctFromHigh > 5) {
    reasons.push("Price dropped >5% from day high — momentum fading");
    score -= 3;
  }

  // --- GREEN CRITERIA ---

  // 1. Meaningful gain (3-12%)
  if (pctChange >= 3 && pctChange <= 12) {
    reasons.push(`Good day gain: +${pctChange.toFixed(1)}%`);
    score += 2;
  } else if (pctChange >= 2) {
    reasons.push(`Moderate gain: +${pctChange.toFixed(1)}%`);
    score += 1;
  } else if (pctChange > 0) {
    reasons.push(`Small gain: +${pctChange.toFixed(1)}%`);
  }

  // 2. Price near day high (buying pressure intact)
  if (pctFromHigh <= 2) {
    reasons.push("Near day high — strong buying pressure");
    score += 2;
  } else if (pctFromHigh <= 4) {
    reasons.push("Slightly off day high");
    score += 1;
  }

  // 3. Volume in top 30% of all stocks
  if (allStocks.length > 0) {
    const volumes = allStocks.map((s) => s.volume).sort((a, b) => b - a);
    const top30Index = Math.floor(volumes.length * 0.3);
    const top30Threshold = volumes[top30Index] || 0;
    if (volume >= top30Threshold && volume > 0) {
      reasons.push("High volume — top 30% of market");
      score += 2;
    }
  }

  // 4. Gap up opening (opened above prev close)
  const gapPct =
    stock.prev_close > 0
      ? ((stock.open - stock.prev_close) / stock.prev_close) * 100
      : 0;
  if (gapPct >= 2) {
    reasons.push(`Gap up opening: +${gapPct.toFixed(1)}%`);
    score += 1;
  }

  // 5. AI score (if available)
  if (aiScore !== undefined) {
    if (aiScore >= 75) {
      reasons.push(`Strong AI score: ${aiScore}`);
      score += 3;
    } else if (aiScore >= 65) {
      reasons.push(`Good AI score: ${aiScore}`);
      score += 2;
    } else if (aiScore >= 50) {
      reasons.push(`Moderate AI score: ${aiScore}`);
      score += 1;
    } else {
      reasons.push(`Low AI score: ${aiScore}`);
      score -= 1;
    }
  }

  // --- DETERMINE SIGNAL ---

  if (score >= 6) {
    return { signal: "GREEN", label: "Strong Buy", reasons };
  }
  if (score >= 4) {
    return { signal: "GREEN", label: "Buy", reasons };
  }
  if (score >= 2) {
    return { signal: "YELLOW", label: "Watch", reasons };
  }
  if (score >= 0) {
    return { signal: "GRAY", label: "Neutral", reasons };
  }
  return { signal: "RED", label: "Avoid", reasons };
}
